Uchu.StandardScripts/Base/GenericActivityManager.cs

```diff
@@ -0,0 +1,381 @@
+        /// <returns>The value of the player.</returns>
+        public float GetActivityValue(Player player, int index)
+        {
+            return this.GetActivityUserData(player, index);
+        }
+        
+        /// <summary>
+        /// Stops the activity for a player.
+        /// </summary>
+        /// <param name="player">Player to stop the activity for.</param>
+        /// <param name="score">Score to finish with.</param>
+        /// <param name="value1">Value 1 to finish with.</param>
+        /// <param name="value2">Value 2 to finish with.</param>
+        /// <param name="quit">Whether the player quit.</param>
+        public void StopActivity(Player player, float score, float value1 = -1, float value2 = -1, bool quit = false)
+        {
+            if (quit)
+            {
+                // Remove the player from the activity.
+                this.RemoveActivityUser(player);
+            }
+            else
+            {
+                // Set the values.
+                this.SetActivityUserData(player, 0, score);
+                if (value1 >= 0)
+                {
+                    this.SetActivityUserData(player, 1, value1);
+                }
+                if (value2 >= 0)
+                {
+                    this.SetActivityUserData(player, 2, value2);
+                }
+                
+                // Distribute the rewards.
+                this.DistributeActivityRewards(player, true, true);
+                
+                // Update the leaderboard.
+                // TODO: self:UpdateActivityLeaderboard{ userID = player }
+                
+                // Send requesting the activity summary.
+                var activityId = this.GetActivityId();
+                player.Message(new RequestActivitySummaryLeaderboardDataMessage()
+                {
+                    Associate = player,
+                    GameId = activityId,
+                    QueryType = QueryType.TopCharacter,
+                });
+                player.Message(new NotifyClientObjectMessage()
+                {
+                    Associate = this.GameObject,
+                    Name = "ToggleLeaderBoard",
+                    Param1 = activityId,
+                    ParamObj = player,
+                });
+                
+                // Remove the player from the activity.
+                this.RemoveActivityUser(player);
+            }
+        }
+
+        /// <summary>
+        /// Sends the RequestActivitySummaryLeaderboardDataMessage message to a client.
+        /// </summary>
+        /// <param name="player">Player to send.</param>
+        /// <param name="activityId">Activity id to send.</param>
+        /// <param name="numberOfResults">Number of results to send.</param>
+        public void GetLeaderboardData(Player player, int activityId, int numberOfResults)
+        {
+            player.Message(new RequestActivitySummaryLeaderboardDataMessage()
```

Comment:
From what I remember, when the client sees, this game message, it echoes the request back to the server. Mick says the request is used, but since we don't send any data back currently as part of global leaderboards, it doesn't do anything.

Reply:
ah okay. Will look at this later when adding leaderboards.